### PR TITLE
Feature/powerups

### PR DESCRIPTION
**Pull request by JordiRomagosa**
Changes made:

- Unified the enemy drops into a single class for light orbs (healing) and powerups.
- Implemented two powerups: increased attack and invulnerability.
- Powerups have a greater chance to drop as more enemies are killed, but the chances are never 0% nor 100%.
- Added little feedback (vfx) for the players to know they have the powerup active.

Still waiting for the UI feedback that is not implemented yet.

### CODE REVIEW

File: Scripts/src/Script/LightOrb.cpp [hunk 1]
```diff
@@ -62,180 +63,64 @@ void LightOrb::Awake()
...
 
-// Use this for initialization
-void LightOrb::Start()
-{
-	player_1 = world->GetPlayer1();
-	player_2 = world->GetPlayer2();
-}
-
 // Update is called once per frame
 void LightOrb::Update()
```
Comment (Dartemiss) [inline Scripts/src/Script/LightOrb.cpp:67]: Why is LightOrb class needed now?

File: Scripts/src/Script/LightOrb.h [hunk 2]
```diff
@@ -12,22 +12,14 @@ class WorldManager;
...
-		TRAVELLING,
-		DEAD
-	};
-
 	LightOrb();
 	~LightOrb() = default;
 
 	void Awake() override;
-	void Start() override;
+	//void Start() override;
```
Comment (Dartemiss) [inline Scripts/src/Script/LightOrb.h:22]: If not used delete it

File: Scripts/src/Script/PlayerController.cpp [hunk 3]
```diff
@@ -633,13 +636,57 @@ bool PlayerController::CheckDashBehaviour()
...
+	time_remaining_attack_powerup = 0;
+	time_remaining_invulnerable_powerup = 0;
+
+	powerup_attack_VFX->Stop();
+	powerup_invul_VFX->Stop();
+}
+
+void PlayerController::UpdatePowerups()
+{
+	float delta_time = App->time->delta_time / 1000;
```
Comment (Dartemiss) [inline Scripts/src/Script/PlayerController.cpp:651]: 1000.f

File: Scripts/src/Script/PlayerController.cpp [hunk 4]
```diff
@@ -633,13 +636,57 @@ bool PlayerController::CheckDashBehaviour()
...
+
+	if (time_remaining_attack_powerup != 0)
```
Comment (Dartemiss) [inline Scripts/src/Script/PlayerController.cpp:653]: .f

File: Scripts/src/Script/PowerUp.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+				{
+					shine_texture_invul = static_cast<ComponentBillboard*>(ch->GetComponent(Component::ComponentType::BILLBOARD));
+				}
+				else if (ch->tag == "exterior_texture_light_orb_VFX")
+				{
+					exterior_texture_invul = static_cast<ComponentBillboard*>(ch->GetComponent(Component::ComponentType::BILLBOARD));
+				}
+			}
+		}
+	}
```
Comment (Dartemiss) [inline Scripts/src/Script/PowerUp.cpp:76]: Probably is more readable to have all of this on a separate method

File: Scripts/src/Script/PowerUp.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	switch (powerup_type)
+	{
+	case PowerUpType::ATTACK:
+		player_to_deliver_controller->time_remaining_attack_powerup += attack_powerup_duration;
+		player_to_deliver_controller->attack_powerup_bonus = 1 + attack_powerup_bonus;
+		break;
+
+	case PowerUpType::INVULNERABLE:
+		player_to_deliver_controller->time_remaining_invulnerable_powerup += invulnerable_powerup_duration;
```
Comment (Dartemiss) [inline Scripts/src/Script/PowerUp.cpp:130]: This means that if a player gets 3 powerups at the same time he will have the bonus the triple of the time right?

File: Scripts/src/Script/PowerUp.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+	case EnemyDropLifeState::WAITING:
+		audio_source->PlayEvent("play_orbs_dropping");
+		break;
+
+	case EnemyDropLifeState::TRAVELLING:
+		break;
+	}
+}
+
+void PowerUp::PickupActivatePowerup()
```
Comment (Dartemiss) [inline Scripts/src/Script/PowerUp.cpp:117]: This method can be const

File: Scripts/src/Script/PowerUp.cpp [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+void PowerUp::Spawn(const float3& spawn_position)
+{
+	EnemyDrop::Spawn(spawn_position);
+	SetupPowerupVfx();
+
+	picked = false;
+}
+
+void PowerUp::SetupPowerupVfx()
```
Comment (Dartemiss) [inline Scripts/src/Script/PowerUp.cpp:143]: Can be cosnt aswell

**Review by Dartemiss — CHANGES_REQUESTED** [review]
GJ man!
I tested the branch and some things:
- The branch is very deprecated you have to merge develop here
- The level1/level2/boss scenes crashes because you just used level1_powerpus.scene for testing, please delete this scene from production
- I think the drops should have a lot more range to get collected like the current lightorbs otherwise is very tedious to get all of them (attracting_radio  should be bigger)
- The damage bonus is good but the invulnerable bonus doesn't have any feedback (or at least i couldn't see anything so I looked like a bug not receiving damage)
- The light ball doesn't have bonus damage?

File: Scripts/src/Script/PowerUp.cpp (hunk 6 above)
Comment (JordiRomagosa) [reply Scripts/src/Script/PowerUp.cpp:130]: Yes, it is. In fact is very hard due to the drop rate, but for now I made it like this. To change it to "refresh duration" is as simple as changing the += for a =.

**Comment by JordiRomagosa** [comment]
Realized requested changes and updated to develop.

**Review by Dartemiss — APPROVED** [review]
Okay looks good now!
Prepare some prefabs of the changes in the scenes. After we merge the big branches with the new level1/level2 and boss we'll merge this branch first after those.

As general comments I think that the red powerup is too long but is just my opinion and can be discussed for next delivery. 
GJ

**Pull request merged**